src/vm/reflectioninvocation.cpp

```diff
@@ -1092,12 +1092,28 @@ FCIMPL5(Object*, RuntimeMethodHandle::InvokeMethod,
     // if we have the magic Value Class return, we need to allocate that class
     // and place a pointer to it on the stack.
 
+    BOOL hasRefReturnAndNeedsBoxing = FALSE; // Indicates that the method has a BYREF return type and the target type needs to be copied into a preallocated boxed object.
+
     TypeHandle retTH = gc.pSig->GetReturnTypeHandle();
+
+    TypeHandle refReturnTargetTH;  // Valid only if retType == ELEMENT_TYPE_BYREF. Caches the TypeHandle of the byref target.
     BOOL fHasRetBuffArg = argit.HasRetBuffArg();
-    CorElementType retType = retTH.GetInternalCorElementType();
-    if (retType == ELEMENT_TYPE_VALUETYPE || fHasRetBuffArg) {
+    CorElementType retType = retTH.GetSignatureCorElementType();
+    BOOL hasValueTypeReturn = retTH.IsValueType() && retType != ELEMENT_TYPE_VOID;
+    if (hasValueTypeReturn || fHasRetBuffArg) {
         gc.retVal = retTH.GetMethodTable()->Allocate();
     }
+    else if (retType == ELEMENT_TYPE_BYREF)
+    {
+        refReturnTargetTH = retTH.AsTypeDesc()->GetTypeParam();
+
+        // If the target of the byref is a value type, we need to preallocate a boxed object to hold the managed return value.
+        if (refReturnTargetTH.IsValueType())
```

Comment:
Can we get here with `ref void`? I know that you cannot type `ref void`, but it may be possible in IL.

Reply:
Yes, it's possible in IL and when I try it, the Invoke succeeds (and calls the method) and gives me a boxed `Void` object. 

I think the runtime loader or JIT should have raised an objection?

Reply:
`GetMethodTable()->Allocate()` does not check whether it is ok to allocate instance the MethodTable. It assumes that the caller verified.

I think similar a problem can happen for `ref Span<char>`.

Reply:
I meant that the "ref void" method itself shouldn't ever have gotten past the JIT phase. "ref Span" is a legal method - just not invokable by a boxing api. Reflection does screen for ref-to-byreflike so we're covered there. We could screen for ref to void in the same place but it doesn't seem like the right place to check that - "ref void" just shouldn't be allowed.

Reply:
This is the endless discussion whether `ref X` is just `X*` with added GC tracking, or whether it has special restrictions attached to it (cannot be null, cannot point to void, ...).

IL treats it like `X*` with added GC tracking. `ref void` was always allowed in IL. I do not see a good reason to introduce a breaking change to disallow it.

C# has special restrictions for it.

This dichotomy has always been a source of confusion.